File: services/gemini_service.py
"""
Gemini Vision Service
---------------------
Called once per dwell event (fire-and-forget, never blocks cards):

  identify_and_update_profile() — Identify product, extract taste signals,
                                   write to Backboard asynchronously.

The profile update feeds the *next* dwell's Cat 2 pipeline, not the current one.
"""

import asyncio
import json
import google.generativeai as genai
from config.settings import settings

genai.configure(api_key=settings.GEMINI_API_KEY)
model = genai.GenerativeModel(settings.GEMINI_MODEL)


# ── Product Identification + Taste Signal Extraction ─────────────────────────

IDENTIFY_PROMPT = """
You are analyzing a screenshot from a social media feed (Instagram or TikTok).

Identify the primary product visible and extract taste signals.
Respond ONLY with valid JSON — no markdown, no explanation.

{{
  "product_name": "short descriptive name",
  "product_category": "e.g. sneakers, hoodie, water bottle",
  "style_signals": ["list", "of", "style", "descriptors"],
  "color_signals": ["list", "of", "colors"],
  "estimated_price_range": "$XX-$XX or 'unknown'",
  "brand_guess": "brand name or 'unknown'",
  "search_query": "best search query to find this product to buy online"
}}

Page URL for context: {page_url}
Page title: {page_title}
"""


async def identify_and_update_profile(
    screenshot_b64: str,
    page_url: str,
    page_title: str,
    user_id: str,
) -> dict:
    """
    Sends viewport screenshot to Gemini Vision, extracts signals,
    and writes them to Backboard. Fire-and-forget — the pipeline
    does not wait on this to return cards.

    Returns the extracted signals dict (for logging only).
    """
    from services import backboard_service

    prompt = IDENTIFY_PROMPT.format(page_url=page_url, page_title=page_title or "unknown")

    try:
        response = await asyncio.to_thread(
            model.generate_content,
            [
                {"mime_type": "image/jpeg", "data": screenshot_b64},
                prompt,
            ],
            request_options={"timeout": 30},
        )

        raw = response.text.strip()
        if raw.startswith("```"):
            raw = raw.split("```")[1]
            if raw.startswith("json"):
                raw = raw[4:]
        signals = json.loads(raw.strip())

    except Exception as e:
        print(f"[Gemini] identify_and_update_profile failed: {e}")
        signals = {
            "product_name": "unknown",
            "product_category": "unknown",
            "style_signals": [],
            "color_signals": [],
            "estimated_price_range": "unknown",
            "brand_guess": "unknown",
            "search_query": "",
        }

    # Write signals to Backboard (non-fatal if it fails)
    try:
        await backboard_service.update_profile(user_id, signals)
    except Exception as e:
        print(f"[Gemini] Backboard write failed: {e}")

    return signals

File: src/services/geminiService.ts
import { GoogleGenerativeAI } from "@google/generative-ai";
import { z } from "zod";

import { settings } from "../config/settings.js";
import * as backboardService from "./backboardService.js";
import type { GeminiSignals } from "./backboardService.js";

const genAI = new GoogleGenerativeAI(settings.GEMINI_API_KEY);
const MODEL_CANDIDATES = [
  "gemini-2.5-flash"
];

let selectedModelName: string | undefined;

const GeminiSignalsSchema = z.object({
  product_name: z.string(),
  product_category: z.string(),
  style_signals: z.array(z.string()),
  color_signals: z.array(z.string()),
  estimated_price_range: z.string(),
  brand_guess: z.string(),
});

const FALLBACK_SIGNALS: GeminiSignals = {
  product_name: "unknown",
  product_category: "unknown",
  style_signals: [],
  color_signals: [],
  estimated_price_range: "unknown",
  brand_guess: "unknown",
};

const IDENTIFY_PROMPT = `You are a shopping assistant analyzing a screenshot from social media (Instagram or TikTok).

Your task: identify the PRIMARY PHYSICAL PRODUCT that a viewer would most likely want to purchase after seeing this post.

Focus on products like:
- Tools, gadgets, or devices being used or demonstrated (e.g. a milk frother, camera, blender)
- Clothing, shoes, or accessories being worn
- Home goods, furniture, or décor being featured
- Beauty or personal care products being applied

Do NOT identify:
- Food or drinks being prepared or consumed — instead identify the TOOL being used to make them
- Abstract moods, activities, or settings
- The background or non-product elements

For style_signals and color_signals, extract short clean adjectives that describe the product's aesthetic (e.g. "minimalist", "matte black", "retro"). Do NOT include parenthetical explanations like "(frother)" or "(drink)" — just the adjective itself.

Respond ONLY with valid JSON, no markdown, no explanation:
{
  "product_name": string,
  "product_category": string,
  "style_signals": string[],
  "color_signals": string[],
  "estimated_price_range": string,
  "brand_guess": string
}
Page URL: {pageUrl}
Page title: {pageTitle}`;

const stripMarkdownFences = (text: string): string => {
  const trimmed = text.trim();
  if (!trimmed.startsWith("```")) {
    return trimmed;
  }

  const noStart = trimmed.replace(/^```(?:json)?\s*/i, "");
  return noStart.replace(/\s*```$/, "").trim();
};

const isMissingModelError = (error: unknown): boolean => {
  if (typeof error !== "object" || error === null) {
    return false;
  }

  const maybeStatus = "status" in error ? (error as { status?: unknown }).status : undefined;
  if (maybeStatus === 404) {
    return true;
  }

  const maybeMessage =
    "message" in error ? (error as { message?: unknown }).message : undefined;
  return typeof maybeMessage === "string" && maybeMessage.toLowerCase().includes("not found");
};

const generateWithModelFallback = async (
  parts: Array<{ inlineData: { mimeType: string; data: string } } | string>,
): Promise<string> => {
  const orderedCandidates = selectedModelName
    ? [selectedModelName, ...MODEL_CANDIDATES.filter((name) => name !== selectedModelName)]
    : MODEL_CANDIDATES;

  let lastError: unknown;

  for (const modelName of orderedCandidates) {
    try {
      const model = genAI.getGenerativeModel({ model: modelName });
      const result = await model.generateContent(parts);

      // Inspect the response before calling .text() — if the response was
      // blocked by a safety filter the SDK sets finishReason to "SAFETY" and
      // leaves candidate.content undefined, causing .text() to throw
      // "Cannot read properties of undefined".
      const candidate = result.response.candidates?.[0];
      const finishReason = candidate?.finishReason ?? "UNKNOWN";
      if (finishReason !== "STOP" && finishReason !== "MAX_TOKENS") {
        const safetyRatings = JSON.stringify(candidate?.safetyRatings ?? []);
        const promptFeedback = JSON.stringify(result.response.promptFeedback ?? {});
        throw new Error(
          `[Gemini] Response blocked or empty — finishReason=${finishReason} safetyRatings=${safetyRatings} promptFeedback=${promptFeedback}`,
        );
      }

      if (selectedModelName !== modelName) {
        console.log(`[Gemini] Using model: ${modelName}`);
      }
      selectedModelName = modelName;
      return result.response.text();
    } catch (error) {
      lastError = error;
      if (isMissingModelError(error)) {
        console.warn(`[Gemini] Model unavailable, trying next: ${modelName}`);
        continue;
      }
      throw error;
    }
  }

  throw lastError ?? new Error("No working Gemini model found");
};

export const identifyAndUpdate = async (
  screenshotB64: string,
  userId: string,
  pageUrl: string,
  pageTitle?: string,
): Promise<void> => {
  let signals: GeminiSignals = FALLBACK_SIGNALS;

  try {
    const imagePart = {
      inlineData: {
        mimeType: "image/jpeg",
        data: screenshotB64,
      },
    };

    const prompt = IDENTIFY_PROMPT.replace("{pageUrl}", pageUrl).replace(
      "{pageTitle}",
      pageTitle ?? "unknown",
    );

    const rawText = await generateWithModelFallback([imagePart, prompt]);
    console.log("[Gemini] Raw response text:", rawText.slice(0, 500));

    const jsonText = stripMarkdownFences(rawText);
    console.log("[Gemini] Stripped JSON text:", jsonText.slice(0, 500));

    const parsedJson: unknown = JSON.parse(jsonText);
    const raw = GeminiSignalsSchema.parse(parsedJson);

    // Strip parenthetical qualifiers like "Black (frother)" → "Black"
    // These leak context words into profile tags and ruin catalog matching.
    const cleanSignal = (s: string) => s.replace(/\s*\([^)]*\)/g, "").trim();
    signals = {
      ...raw,
      style_signals: raw.style_signals.map(cleanSignal).filter(Boolean),
      color_signals: raw.color_signals.map(cleanSignal).filter(Boolean),
    };
    console.log("[Gemini] Parsed signals:", JSON.stringify(signals));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    const stack = error instanceof Error ? (error.stack ?? "") : "";
    console.warn("[Gemini] identifyAndUpdate failed:", message);
    if (stack) console.warn("[Gemini] Stack:", stack.split("\n").slice(0, 4).join(" | "));
  }

  try {
    await backboardService.updateProfile(userId, signals);
  } catch (error) {
    console.warn("[Gemini] Backboard update failed", error);
  }
};
